rfcs/20230118-f16-accumulation/README.md

```diff
@@ -0,0 +1,125 @@
+Proposal for setting accumulation datatype.
+
+# 1. Introduction
+
+In order to maintain good accuracy in DNN workloads, it is important
+to use wide-enough accumulators during computation.  The default
+accumulation datatype in oneDNN is f32, for floating point
+computations, and s32 for integer computation.
+
+Some platforms allow additional speedups when using smaller
+accumulation datatype.  This is in particular the case for f16
+primitives, where up to 2x speedups can be obtained by using f16
+accumulation datatype on some platforms.
+
+# 2. Problem statement
+
+As of now, oneDNN officially supports only inference for f16 models,
+and many data show that good levels of accuracy can be maintained
+while using f16 accumulation datatype for most f16 models. We hence
+have been dispatching implementations with f16 accumulation on some
+platforms.
+
+Even though dispatching implementations with f16 accumulation is
+acceptable for inference, it is not for training, which requires the
+full f32 precision accumulators. Hence, as oneDNN is planning
+support for f16 training, it is important that we dispatch only
+implementation with f32 accumulation for training workloads.
+
+Note that the main user for oneDNN f16 implementations is openVINO,
+which is the requester for the implementations with f16 accumulation
+in the first place.
```

Comment:
Currently CPU[x64] only supports f32 as accumulation data type. Is there a request to have f16 accumulator on CPU?... or is it practically only for GPU's?

Reply:
The only reason to do f16 accumulation for inference is when HW does f16 accumulation faster than f32.
This seems to be the case for some Intel GPUs and ARM chips.
This does not seem to be valid for Intel CPUs though.

Reply:
SPR has f16 instructions that accumulate to f16, instead we chose the path `f16 -> f32 -> f32-math -> f16`.
I think, I understand your point.... this RFC is probably not for x64.

Reply:
>This does not seem to be valid for Intel CPUs though.

SPR has f16 instructions that accumulate to f16. Will f16 accumulation be used for SPR ? It should have benefits on SPR.

Reply:
Any ref doc about the requirement of fp32 acc for training? Please note that fp16 acc is default for fp16 data type in PyTorch NV GPU path. It is not specific for inference only.
https://pytorch.org/docs/stable/notes/cuda.html#reduced-precision-reduction-in-fp16-gemms

Reply:
> This does not seem to be valid for Intel CPUs though.
> SPR has f16 instructions that accumulate to f16, instead we chose the path f16 -> f32 -> f32-math -> f16.

It also helps SPR CPU.

Reply:
There is no debate it helps to keep accumulators in reduced precision (especially when we parallelize over the reduction dimension).

Here the question is what should be the default for oneDNN (aka what is safe to do):
- cudnn by default disables f16 reduction ([ref](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnMathType_t)).
-  cublas does not use f16 for accumulation, but for intermediate stores only ([ref](https://docs.nvidia.com/cuda/cublas/#cublascomputetype-t)). This is the same definition as pytorch
- all training experiments we are familiar with rely on f32 accumulation.

This is why, in order to err on the side of caution, I would recommend to use f32 as default accumulation datatype, and allow oneDNN user to enable f16 accumulation when they know f16 accumulation is safe.

Reply:
> cublas does not use f16 for accumulation, but for intermediate stores only ([ref](https://docs.nvidia.com/cuda/cublas/#cublascomputetype-t)). This is the same definition as pytorch

It means CUDA FP16 GEMM mainly use fp32 for accumulation and fp16 for some intermediate results. Can oneDNN also choose to do like this to strike a balance between accuracy and performance？

Reply:
One issue I see is that it could lead to unpredictable behavior if we enable that by default, as the number of intermediate rounding would be different on different HW target.

We could expose an f16_store datatype, but then it would still be confusing I believe, as it provides no accuracy properties on the implementation (the number of intermediate stores is implementation defined).
@petercad @echeresh , how much performance upside would you expect from having such a mode (accumulation in f32 registers, but partial accumulators in f16 in memory)?

Reply:
@mgouicem For GEMM I don't expect much performance advantage from this mode on current GPUs with current GEMM algorithms. Future GPUs though may have support for f16 global atomic addition, in which case this mode would provide a significant performance boost, as much as 50% in some cases.